src/Features/Core/Portable/ConvertForToForEach/AbstractConvertForToForEachCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,504 @@
+                    SyntaxGenerator.DefaultRemoveOptions | SyntaxRemoveOptions.KeepLeadingTrivia);
+            }
+
+            var options = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+            editor.ReplaceNode(
+                forStatement,
+                (currentFor, _) => this.ConvertForNode(
+                    (TForStatementSyntax)currentFor, typeNode, foreachIdentifier,
+                    collectionExpression, iterationType, options));
+
+            return document.WithSyntaxRoot(editor.GetChangedRoot());
+
+            // local functions
+            (TTypeNode, SyntaxToken, TStatementSyntax) tryDeconstructInitialDeclaration()
+            {
+                var bodyStatements = GetBodyStatements(forStatement);
+
+                if (bodyStatements.Count >= 1)
+                {
+                    var firstStatement = bodyStatements[0];
+                    if (syntaxFacts.IsLocalDeclarationStatement(firstStatement))
+                    {
+                        var variables = syntaxFacts.GetVariablesOfLocalDeclarationStatement(firstStatement);
+                        if (variables.Count == 1)
+                        {
+                            var firstVariable = (TVariableDeclaratorSyntax)variables[0];
+                            if (IsValidVariableDeclarator(firstVariable))
+                            {
+                                var firstVariableInitializer = syntaxFacts.GetValueOfEqualsValueClause(
+                                    syntaxFacts.GetInitializerOfVariableDeclarator(firstVariable));
+                                if (syntaxFacts.AreEquivalent(firstVariableInitializer, indexExpression))
+                                {
+                                    var type = (TTypeNode)syntaxFacts.GetTypeOfVariableDeclarator(firstVariable)?.WithoutLeadingTrivia();
+                                    var identifier = syntaxFacts.GetIdentifierOfVariableDeclarator(firstVariable);
+                                    var statement = firstStatement;
+                                    return (type, identifier, statement);
+                                }
+                            }
+                        }
+                    }
+                }
+
+                return default;
+            }
+
+            void findAndReplaceMatches(SyntaxNode current)
+            {
+                if (syntaxFacts.AreEquivalent(current, indexExpression))
+                {
+                    // Found a match.  replace with iteration variable.
+                    var replacementToken = foreachIdentifierReference;
+
+                    if (semanticFacts.IsWrittenTo(semanticModel, current, cancellationToken))
+                    {
+                        replacementToken = replacementToken.WithAdditionalAnnotations(
+                            WarningAnnotation.Create(FeaturesResources.Warning_colon_Collection_was_modified_during_iteration));
```

Comment:
Ah, this is covered. Does `IsWrittenTo` mark any invocation on collection as a write? If so, ignore one of my previous comments about collection being modified.